app/lib/analyzer.dart

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library pub_dartlang_org.analysis;
+
+import 'dart:async';
+import 'dart:io';
+
+import 'package:yaml/yaml.dart' as yaml;
+
+import 'model_properties.dart';
+import 'utils.dart';
+
+/// Collects the preparation, analysis and cleanup functions for analyzing a
+/// package.
+class Analyzer {
+  final File _sourceTarGz;
+  final Directory _workDir;
+  final bool _clearWorkDir;
+  Directory _packageDir;
+
+  Pubspec pubspec;
+  PubspecLock pubspecLock;
+
+  Analyzer._(this._sourceTarGz, this._workDir, this._clearWorkDir);
+
+  factory Analyzer(String source, {String workDir}) {
+    final Directory wd = workDir != null
+        ? new Directory(workDir)
+        : Directory.systemTemp.createTempSync();
+    return new Analyzer._(new File(source), wd, workDir == null);
+  }
```

Comment:
Maybe it would be easier to read when using only one instead of two constructors:

```dart
Analyzer(String source, {String workdir}) 
    : _sourceTarGz = new File(source),
      _workDir  = workdir != null ? new Directory(workdir) : Directory.systemTemp.createTempSync(),
      _clearWorkDir = workdir == null;
```